compiler/luci/pass/src/FoldReshapePass.test.cpp

```diff
@@ -0,0 +1,197 @@
+
+namespace
+{
+
+std::unique_ptr<luci::CircleQuantParam> gen_qparam(const std::vector<float> &s,
+                                                   const std::vector<int64_t> &zp)
+{
+  auto qparam = std::make_unique<luci::CircleQuantParam>();
+  {
+    for (auto scale : s)
+      qparam->scale.push_back(scale);
+
+    for (auto zerop : zp)
+      qparam->zerop.push_back(zerop);
+  }
+
+  return std::move(qparam);
+}
+
+template <loco::DataType DT> class FoldReshapeTest : public luci::ConstantFoldingAddTestGraph
+{
+public:
+  FoldReshapeTest(std::initializer_list<uint32_t> input_shape,
+                  std::initializer_list<uint32_t> output_shape)
+    : luci::ConstantFoldingAddTestGraph(output_shape, DT)
+  {
+    _reshape = _g.nodes()->template create<luci::CircleReshape>();
+    _x = _g.nodes()->template create<luci::CircleConst>();
+    _shape = _g.nodes()->template create<luci::CircleConst>();
+
+    _reshape->dtype(DT);
+    _x->dtype(DT);
+    _shape->dtype(loco::DataType::S32);
+
+    _reshape->shape(_shape);
+    _x->shape(input_shape);
+    _shape->shape({static_cast<uint32_t>(output_shape.size())});
+
+    uint32_t num_elems = 1;
+    for (auto dim : input_shape)
+      num_elems *= dim;
+
+    _x->size<DT>(num_elems);
+    for (uint32_t i = 0; i < num_elems; i++)
+      _x->at<DT>(i) = i;
+
+    _shape->size<loco::DataType::S32>(output_shape.size());
+    uint32_t i = 0;
+    for (auto dim : output_shape)
+    {
+      _shape->at<loco::DataType::S32>(i++) = static_cast<int32_t>(dim);
+    }
+
+    if (DT == loco::DataType::S16)
+    {
+      _x->quantparam(gen_qparam({1.0}, {0}));
+      _reshape->quantparam(gen_qparam({1.0}, {0}));
+    }
+
+    _reshape->tensor(_x);
+    _reshape->shape(_shape);
+
+    _reshape->name("reshape");
+    _shape->name("shape");
+    _x->name("x");
+  }
+
+  loco::Node *createFoldedPattern() override { return _reshape; }
+
+public:
+  void set_unknown_dim() { _x->dim(0).unset(); }
+  void set_non_per_tensor() { _x->quantparam(gen_qparam({1.0, 2.0}, {0, 0})); }
+
+protected:
+  luci::CircleReshape *_reshape = nullptr;
+  luci::CircleConst *_x = nullptr;
+  luci::CircleConst *_shape = nullptr;
+};
+
+/**
+ *  Graph that has a Reshape Op with constant input
+ *
+ *    BEFORE
+ *
+ *         [CircleConst]
+ *               |
+ *            [Reshape]
+ *
+ *    AFTER
+ *
+ *         [CircleConst]
+ *
+ */
+class FoldFP32ReshapeTest : public FoldReshapeTest<loco::DataType::FLOAT32>, public ::testing::Test
+{
+public:
+  FoldFP32ReshapeTest() : FoldReshapeTest<loco::DataType::FLOAT32>({1, 3}, {3}) {}
+
+  virtual void SetUp() { init(); }
+};
+
+class FoldS16ReshapeTest : public FoldReshapeTest<loco::DataType::S16>, public ::testing::Test
+{
+public:
+  FoldS16ReshapeTest() : FoldReshapeTest<loco::DataType::S16>({1, 3}, {3}) {}
+
+  virtual void SetUp() { init(); }
+};
+
+} // namespace
+
+TEST_F(FoldFP32ReshapeTest, fold_reshape_fp32)
+{
+  luci::FoldReshapePass pass;
+  while (pass.run(graph()))
+    ;
```

Comment:
@jinevening 

I don't know it well, May I ask it?
Is there any reason to use `infinite while~` grammar?

Reply:
As talked offline, pass is typically applied iteratively until graph is not changed. This code simulates such behavior.